qdao/circuit.py

```diff
@@ -109,7 +112,99 @@ def run(self, circuit: Any) -> List[QdaoCircuit]:
         if instrs:
             sub_circ = self._circ_helper.gen_sub_circ(instrs, self._nl, self._np)
             sub_circs.append(sub_circ)
+        print("----------StaticPartitioner-----------")
+        print("num of sub-circuits:" + str(len(sub_circs)))
+        return sub_circs
+
+
+class dptask:
+    """To assist the Uniq partitioning algorithm"""
+
+    def __init__(self, qubitNum: int, gateNum: int):
+        self.gateNum = gateNum
+        self.qubitNum = qubitNum
+
+    def addGate(self, gateIndex: int, gatePos):
+        if isinstance(gatePos, int):
+            gatePos = [gatePos]
+        for j in range(self.qubitNum):
+            self.result_bit[gateIndex + 1][j] += self.result_bit[gateIndex][j]
+            self.result_op[gateIndex + 1][j] += self.result_op[gateIndex][j]
+            if j in gatePos:
+                self.result_op[gateIndex + 1][j].append(gateIndex + 1)
+                if len(gatePos) == 2:
+                    a = [x for x in gatePos if x != j][0]
+                    self.result_bit[gateIndex + 1][j] += self.result_bit[gateIndex][a]
+                    self.result_bit[gateIndex + 1][j] = list(
+                        set(self.result_bit[gateIndex + 1][j])
+                    )
+                    self.result_op[gateIndex + 1][j] += self.result_op[gateIndex][a]
+                    self.result_op[gateIndex + 1][j] = list(
+                        set(self.result_op[gateIndex + 1][j])
+                    )
+
+    def createTask(self, ops):
+        self.result_bit = [
+            [[] for _ in range(self.qubitNum)] for _ in range(self.gateNum + 1)
+        ]
+        self.result_op = [
+            [[] for _ in range(self.qubitNum)] for _ in range(self.gateNum + 1)
+        ]
+        for i in range(self.qubitNum):
+            self.result_bit[0][i].append(i)
+        for i in range(self.gateNum):
+            self.addGate(i, ops[i].pos)
+
+    def selectSubCircuit(self, numSubcircuit: int) -> (list[int], int):
+        """Find sub-lines that meet the requirements from the array in
+        preprocessing. The return value is the number of bits representing
+        the sub-circuit. When a sub-circuit that fails to obtain enough
+        qubits is selected at one time, the return value can be used for
+        the next selection."""
+        numQubit = 0
+        numOp = 0
+        opList = []
+        bitList = []
+        for i in range(self.gateNum):
+            for j in range(self.qubitNum):
+                if (
+                    len(self.result_op[i + 1][j]) >= numOp
+                    and len(self.result_bit[i + 1][j]) <= numSubcircuit
+                ):
+                    opList = self.result_op[i + 1][j]
+                    bitList = self.result_bit[i + 1][j]
+                    numQubit = len(self.result_bit[i + 1][j])
+                    numOp = len(self.result_op[i + 1][j])
+        return (opList, numQubit)
+
+
+class UniQPartitioner(BasePartitioner):
```

Comment:
Maybe we should consider registering `UniQPartitioner` into [PARTITIONERS](https://github.com/Zhaoyilunnn/qdao/pull/16/files#diff-41c4e231f629bf43032e3b73b55a7942e730c6e096f28cb22d7e64724e20649fR211). This is somehow a naive factory design pattern offering better maintainablility. We could discuss more elegant practice.